# Performance Stats — Eduardo Palmieri Dataset

Objective: load individual player performance statistics for Série A 2024
from a secondary source, and prepare it to be merged (by player name) with
df_players_brazil from notebook 02.

In [197]:
import pandas as pd

## Loading performance stats data

In [198]:
df_stats = pd.read_csv('../data/raw/database.csv')

## Exploring structure

In [199]:
df_stats.head()

,Jogador,Time,#,Nação,Pos.,Idade,Min.,Gols,Assis.,PB,...,GCA,Cmp,Att,Cmp%,PrgP,Conduções,PrgC,Tent,Suc,Data
0,Éder,Criciúma,23,ITA,FW,37-150,57.0,0,0,0,...,0.0,12.0,16.0,"75,0",3.0,17.0,0.0,0.0,0.0,2024-04-13
1,Yannick Bolasie,Criciúma,11,COD,FW,34-325,33.0,0,0,0,...,0.0,4.0,5.0,"80,0",0.0,6.0,3.0,5.0,3.0,2024-04-13
2,Renato Kayser,Criciúma,79,BRA,FW,28-056,85.0,1,0,0,...,0.0,10.0,15.0,"66,7",4.0,14.0,0.0,1.0,0.0,2024-04-13
3,Arthur,Criciúma,35,BRA,FW,31-303,5.0,0,0,0,...,0.0,5.0,6.0,"83,3",1.0,2.0,0.0,1.0,1.0,2024-04-13
4,Marquinhos Gabriel,Criciúma,10,BRA,AM,33-267,77.0,0,0,0,...,1.0,22.0,24.0,"91,7",3.0,24.0,2.0,7.0,5.0,2024-04-13


In [200]:
df_stats.columns

Index(['Jogador', 'Time', '#', 'Nação', 'Pos.', 'Idade', 'Min.', 'Gols',
       'Assis.', 'PB', 'PT', 'TC', 'CaG', 'CrtsA', 'CrtV', 'Contatos', 'Div',
       'Crts', 'Bloqueios', 'xG', 'npxG', 'xAG', 'SCA', 'GCA', 'Cmp', 'Att',
       'Cmp%', 'PrgP', 'Conduções', 'PrgC', 'Tent', 'Suc', 'Data'],
      dtype='str')

In [201]:
df_stats.info()

<class 'pandas.DataFrame'>
RangeIndex: 11682 entries, 0 to 11681
Data columns (total 33 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   Jogador    11682 non-null  str    
 1   Time       11682 non-null  str    
 2   #          11682 non-null  int64  
 3   Nação      11676 non-null  str    
 4   Pos.       11682 non-null  str    
 5   Idade      11676 non-null  str    
 6   Min.       11681 non-null  float64
 7   Gols       11682 non-null  int64  
 8   Assis.     11682 non-null  int64  
 9   PB         11682 non-null  int64  
 10  PT         11682 non-null  int64  
 11  TC         11682 non-null  int64  
 12  CaG        11682 non-null  int64  
 13  CrtsA      11682 non-null  int64  
 14  CrtV       11682 non-null  int64  
 15  Contatos   11681 non-null  float64
 16  Div        11681 non-null  float64
 17  Crts       11682 non-null  int64  
 18  Bloqueios  11681 non-null  float64
 19  xG         11681 non-null  float64
 20  npxG       11681 

## Checking date coverage

In [202]:
df_stats['Data'].min()

'2024-04-13'

In [203]:
df_stats['Data'].max()

'2024-12-08'

## Aggregating stats by player

In [204]:
df_stats_by_player = df_stats.groupby('Jogador').agg({
    'Pos.': 'first',      # mantém a posição (assume a primeira registrada)
    'Time': 'first',      # mantém o time (assume o primeiro registrado)
    'Min.': 'sum',
    'Gols': 'sum',
    'Assis.': 'sum',
    'xG': 'sum',
    'xAG': 'sum',
    'PrgP': 'sum',
    'PrgC': 'sum',
    'Tent': 'sum',
    'Suc': 'sum',
    'SCA': 'sum',
    'GCA': 'sum',
    'Div': 'sum',
    'Bloqueios': 'sum',
    'CrtsA': 'sum',
    'CrtV': 'sum',
})
df_stats_by_player

,Pos.,Time,Min.,Gols,Assis.,xG,xAG,PrgP,PrgC,Tent,Suc,SCA,GCA,Div,Bloqueios,CrtsA,CrtV
Jogador,,,,,,,,,,,,,,,,,
Abner,CM,Juventude,108.0,0,0,0.0,0.0,2.0,0.0,0.0,0.0,1.0,0.0,2.0,1.0,1,0
Ademir Santos,LW,Bahia,1131.0,3,4,3.3,2.7,29.0,65.0,62.0,28.0,47.0,8.0,30.0,10.0,0,0
Adriano Martins,CB,Atlético Goianiense,2990.0,0,0,1.2,0.0,106.0,24.0,14.0,8.0,22.0,1.0,43.0,41.0,7,0
Adriel,GK,Bahia,334.0,0,0,0.0,0.0,1.0,0.0,1.0,1.0,0.0,0.0,1.0,0.0,0,0
Adryelson,CB,Botafogo (RJ),580.0,1,0,0.2,0.3,22.0,2.0,0.0,0.0,8.0,0.0,4.0,5.0,0,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
Álvaro Barreal,LW,Cruzeiro,1488.0,2,4,2.3,1.5,69.0,29.0,38.0,17.0,56.0,5.0,38.0,39.0,6,0
Ángel Romero,RW,Corinthians,1449.0,5,4,5.4,2.7,53.0,34.0,49.0,24.0,53.0,7.0,13.0,15.0,6,0
Éder,FW,Criciúma,764.0,0,0,0.9,1.1,22.0,13.0,10.0,1.0,22.0,2.0,7.0,2.0,5,0


## Preparing for merge

In [205]:
df_stats_by_player = df_stats_by_player.reset_index()

In [206]:
df_stats_by_player.columns

Index(['Jogador', 'Pos.', 'Time', 'Min.', 'Gols', 'Assis.', 'xG', 'xAG',
       'PrgP', 'PrgC', 'Tent', 'Suc', 'SCA', 'GCA', 'Div', 'Bloqueios',
       'CrtsA', 'CrtV'],
      dtype='str')

In [207]:
df_players_brazil = pd.read_csv('../data/processed/players_brazil_2024.csv')

In [208]:
df_players_brazil['name_x'].head(10)

0              Fábio
1               Nenê
2       Thiago Silva
3      Dimitri Payet
4    Gabriel Mercado
5         David Luiz
6             Fagner
7             Cássio
8           Fernando
9            Willian
Name: name_x, dtype: str

In [209]:
df_stats_by_player['Jogador'].head(10)

0                Abner
1        Ademir Santos
2      Adriano Martins
3               Adriel
4            Adryelson
5                Adson
6     Agustín Canobbio
7         Agustín Giay
8    Agustín Marchesín
9        Agustín Rogel
Name: Jogador, dtype: str

## Attempting direct name match

In [210]:
df_players_brazil['name_x'].isin(df_stats_by_player['Jogador']).sum()

np.int64(386)

In [211]:
names_without_match = df_players_brazil[~df_players_brazil['name_x'].isin(df_stats_by_player['Jogador'])]
names_without_match['name_x'].head(20)

13         David Braz
16         Léo Vieira
19             Neymar
21       Rafael Tolói
22              Souza
25              Denis
27       Tomás Rincón
34              Oscar
37                Gil
38            Aderlan
42            Everson
43               Vina
44            Eduardo
45               Neto
46    Sérgio Oliveira
47      Cédric Soares
51       Willian José
52     Zakaria Labyad
53    Christian Ortíz
56      Thiago Mendes
Name: name_x, dtype: str

In [212]:
df_stats_by_player[df_stats_by_player['Jogador'].str.contains('Neymar', case=False, na=False)]

,Jogador,Pos.,Time,Min.,Gols,Assis.,xG,xAG,PrgP,PrgC,Tent,Suc,SCA,GCA,Div,Bloqueios,CrtsA,CrtV


In [213]:
df_stats_by_player[df_stats_by_player['Jogador'].str.contains('Vieira', case=False, na=False)]

,Jogador,Pos.,Time,Min.,Gols,Assis.,xG,xAG,PrgP,PrgC,Tent,Suc,SCA,GCA,Div,Bloqueios,CrtsA,CrtV
125,Dudu Vieira,DM,Juventude,268.0,0,0,0.0,0.0,8.0,1.0,3.0,2.0,2.0,0.0,4.0,6.0,1,0
174,Felipe Vieira,CB,Fluminense,161.0,0,0,0.0,0.0,1.0,3.0,2.0,1.0,0.0,0.0,5.0,1.0,0,0


In [214]:
df_stats_by_player[df_stats_by_player['Jogador'].str.contains('verson', case=False, na=False)]

,Jogador,Pos.,Time,Min.,Gols,Assis.,xG,xAG,PrgP,PrgC,Tent,Suc,SCA,GCA,Div,Bloqueios,CrtsA,CrtV
113,Deyverson,FW,Cuiabá,1147.0,3,1,2.9,0.6,19.0,6.0,16.0,6.0,14.0,4.0,6.0,9.0,4,0
650,Éverson,GK,Atlético Mineiro,2152.0,0,0,0.0,0.0,2.0,0.0,0.0,0.0,2.0,1.0,1.0,0.0,5,1


## Normalizing names for matching

In [215]:
import unicodedata

def normalize_name(name):
    if pd.isna(name):
        return name
    name = unicodedata.normalize('NFKD', name)  # separate letter + accent
    name = ''.join(c for c in name if not unicodedata.combining(c))  # remove only the accents
    return name.lower().strip()  # make everything lowercase, no spaces at the ends

In [216]:
df_players_brazil['name_normalized'] = df_players_brazil['name_x'].apply(normalize_name)

In [217]:
df_stats_by_player['name_normalized'] = df_stats_by_player['Jogador'].apply(normalize_name)

In [218]:
df_players_brazil['name_normalized'].isin(df_stats_by_player['name_normalized']).sum()

np.int64(401)

## Investigating remaining mismatches

In [219]:
no_match = df_players_brazil[~df_players_brazil['name_normalized'].isin(df_stats_by_player['name_normalized'])]
no_match['name_x'].head(20)

13         David Braz
16         Léo Vieira
19             Neymar
21       Rafael Tolói
22              Souza
25              Denis
27       Tomás Rincón
34              Oscar
37                Gil
38            Aderlan
43               Vina
44            Eduardo
45               Neto
46    Sérgio Oliveira
47      Cédric Soares
51       Willian José
52     Zakaria Labyad
53    Christian Ortíz
56      Thiago Mendes
57             Marçal
Name: name_x, dtype: str

## Evaluating match rate by club

In [220]:
df_players_brazil['matched'] = df_players_brazil['name_normalized'].isin(df_stats_by_player['name_normalized'])
df_players_brazil.groupby('current_club_name')['matched'].mean()

current_club_name
Ceará Sporting Club                 0.206897
Club de Regatas Vasco da Gama       0.380952
Clube Atlético Mineiro              0.387097
Clube de Regatas do Flamengo        0.423729
Cruzeiro Esporte Clube              0.522727
Esporte Clube Bahia                 0.343750
Esporte Clube Juventude             0.350000
Esporte Clube Vitória               0.475000
Fluminense Football Club            0.492308
Fortaleza Esporte Clube             0.275862
Grêmio Foot-Ball Porto Alegrense    0.380000
Mirasol Futebol Clube               0.316667
Red Bull Bragantino                 0.423077
S. A. F. Botafogo                   0.347826
Santos Futebol Clube                0.224490
Sociedade Esportiva Palmeiras       0.400000
Sport Club Corinthians Paulista     0.488372
Sport Club Internacional            0.541667
Sport Club do Recife                0.200000
São Paulo Futebol Clube             0.333333
Name: matched, dtype: float64

## Manual investigation of remaining mismatches

In [221]:
name_investigate = no_match['name_x'].tolist()

for name in name_investigate:
    last_name = name.split()[-1]  # takes the last word of the name 
    similiar = df_stats_by_player[df_stats_by_player['Jogador'].str.contains(last_name, case=False, na=False)]
    if not similiar.empty:
        print(f"'{name}' → possível correspondência: {similiar['Jogador'].tolist()}")

'Léo Vieira' → possível correspondência: ['Dudu Vieira', 'Felipe Vieira']
'Souza' → possível correspondência: ['Cauly Oliveira Souza', 'Gabriel Souza', 'Hugo Souza', 'Igor Vinícius de Souza', 'Josef de Souza']
'Oscar' → possível correspondência: ['Gabriel Moscardo', 'Oscar Estupiñan']
'Gil' → possível correspondência: ['Gilberto']
'Eduardo' → possível correspondência: ['Carlos Eduardo', 'Eduardo Melo', 'Eduardo Santos', 'Eduardo Sasha', 'Eduardo Vargas', 'Eduardo Vinícius', 'Luis Eduardo Marques dos Santos']
'Sérgio Oliveira' → possível correspondência: ['Cauly Oliveira Souza']
'Cédric Soares' → possível correspondência: ['Everton Soares', 'Fábio Soares', 'Tiquinho Soares']
'Willian José' → possível correspondência: ['José Andrés Martínez', 'José Breno', 'José Cifuentes', 'José Hugo', 'José Hurtado', 'José Luis Rodríguez Bebanz', 'José López', 'José Sabino', 'José Vitor Silva Neves', 'José Welison']
'Thiago Mendes' → possível correspondência: ['Douglas Mendes', 'Matheus Mendes', 'Natha

## Fuzzy matching with club as tie-breaker

In [222]:
from rapidfuzz import fuzz

def find_best_match(nome, clube, stats_df):
    # filters candidates from the same club (comparison 'contains', both ways,
    # because one club name might be an abbreviation of the other)
    candidates = stats_df[stats_df['Time'].apply(
        lambda t: t.lower() in clube.lower() or clube.lower() in t.lower()
    )]
    
    if candidates.empty:
        return None, 0

    # calculates the similarity score between 'name' and each candidate from the same club
    scores = candidates['Jogador'].apply(lambda c: fuzz.WRatio(nome, c))
    best_idx = scores.idxmax()
    best_score = scores.max()
    
    return candidates.loc[best_idx, 'Jogador'], best_score

In [223]:
results = no_match.apply(
    lambda row: find_best_match(row['name_x'], row['current_club_name'], df_stats_by_player),
    axis=1
)

no_match['fuzzy_candidate'] = results.apply(lambda r: r[0])
no_match['fuzzy_score'] = results.apply(lambda r: r[1])

In [224]:
no_match[['name_x', 'current_club_name', 'fuzzy_candidate', 'fuzzy_score']].sort_values('fuzzy_score', ascending=False)

,name_x,current_club_name,fuzzy_candidate,fuzzy_score
265,José Luis Rodríguez,Club de Regatas Vasco da Gama,José Luis Rodríguez Bebanz,95.000000
568,Gustavo Martins,Grêmio Foot-Ball Porto Alegrense,Gustavo Martins Santos,95.000000
460,Ewerthon,Esporte Clube Juventude,Ewerton,93.333333
22,Souza,Club de Regatas Vasco da Gama,Gabriel Souza,90.000000
910,Kaique Kenji,Cruzeiro Esporte Clube,Kaique Kenji Takamura Correa,90.000000
...,...,...,...,...
988,Cauã Berrío,Sport Club do Recife,NaN,0.000000
987,Pedro Esli,Ceará Sporting Club,NaN,0.000000
140,Romarinho,Sport Club do Recife,NaN,0.000000
745,Riquelmy,Sport Club do Recife,NaN,0.000000


In [225]:
no_match['fuzzy_score'].value_counts(bins=[0, 50, 70, 80, 90, 100])

(-0.001, 50.0]    314
(50.0, 70.0]      244
(80.0, 90.0]       48
(70.0, 80.0]       37
(90.0, 100.0]       3
Name: count, dtype: int64

In [226]:
no_match[(no_match['fuzzy_score'] >= 70) & (no_match['fuzzy_score'] < 90)][['name_x', 'current_club_name', 'fuzzy_candidate', 'fuzzy_score']].sort_values('fuzzy_score', ascending=False)

,name_x,current_club_name,fuzzy_candidate,fuzzy_score
764,Isac,Fluminense Football Club,Isaac,88.888889
462,Luan Freitas,Esporte Clube Juventude,Lucas Freitas,88.000000
548,José Manuel López,Sociedade Esportiva Palmeiras,José López,85.500000
566,Rodrigo Santos,Fortaleza Esporte Clube,Luis Eduardo Marques dos Santos,85.500000
575,Ignacio Sosa,Red Bull Bragantino,Ignacio Laquintana,85.500000
567,Mateus Silva,Esporte Clube Vitória,Mateus Gonçalves Martins,85.500000
466,José Andrés Hurtado,Red Bull Bragantino,José Hurtado,85.500000
296,Gabriel Rojas,Cruzeiro Esporte Clube,Tevis Gabriel Alves Santos,85.500000
648,Phillipe Gabriel,Club de Regatas Vasco da Gama,Zé Gabriel,85.500000
356,Caíque Gonçalves,Esporte Clube Vitória,Mateus Gonçalves Martins,85.500000


In [227]:
no_match[no_match['name_x'].isin(['Léo Linck', 'Mosquito'])][['name_x', 'current_club_name', 'fuzzy_candidate', 'fuzzy_score']]

,name_x,current_club_name,fuzzy_candidate,fuzzy_score
504,Léo Linck,S. A. F. Botafogo,NaN,0.0
1037,Mosquito,Clube Atlético Mineiro,Júnior Alonso,45.0


### Issue found: duplicate candidate assignment

Reviewing the results above revealed that the same candidate could be
assigned as the "best match" for multiple different players (e.g. three
different names all matched to "Tevis Gabriel Alves Santos" with the same
score). This happens because each row is matched independently, with no
check for whether a candidate was already used. Since a real person can
only be one player, this is a genuine one-to-one matching problem —
addressed in the next section.

## One-to-one fuzzy matching with club filter (corrected approach)

In [228]:
from rapidfuzz import fuzz

# exclude candidates already used in the direct/normalized match (the 401)
already_matched_names = df_players_brazil[
    df_players_brazil['name_normalized'].isin(df_stats_by_player['name_normalized'])
]['name_normalized']

df_stats_available = df_stats_by_player[~df_stats_by_player['name_normalized'].isin(already_matched_names)]

# 1. calculate all possible pairs (unmatched player x candidate from the same club)
pairs = []
for idx, row in no_match.iterrows():
    name = row['name_x']
    club = row['current_club_name']
    candidates = df_stats_available[df_stats_available['Time'].apply(
        lambda t: t.lower() in club.lower() or club.lower() in t.lower()
    )]
    for _, cand_row in candidates.iterrows():
        score = fuzz.WRatio(name, cand_row['Jogador'])
        pairs.append({
            'player_idx': idx,
            'name_x': name,
            'candidate': cand_row['Jogador'],
            'score': score
        })

df_pairs = pd.DataFrame(pairs)

In [229]:
# 2. sort by score, highest first
df_pairs = df_pairs.sort_values('score', ascending=False)

# 3. go through in order, only accepting a match if neither the player
# nor the candidate has been used yet
used_players = set()
used_candidates = set()
accepted_matches = []

for _, pair in df_pairs.iterrows():
    if pair['score'] < 81:
        break  # already past the threshold, stop looking
    if pair['player_idx'] in used_players or pair['candidate'] in used_candidates:
        continue  # player or candidate already matched, skip this pair
    accepted_matches.append(pair)
    used_players.add(pair['player_idx'])
    used_candidates.add(pair['candidate'])

df_accepted_matches = pd.DataFrame(accepted_matches)
df_accepted_matches

,player_idx,name_x,candidate,score
2524,568,Gustavo Martins,Gustavo Martins Santos,95.000000
799,265,José Luis Rodríguez,José Luis Rodríguez Bebanz,95.000000
1876,460,Ewerthon,Ewerton,93.333333
1334,360,Eduardo,Eduardo Santos,90.000000
380,144,João Ricardo,João Ricardo Riedi,90.000000
533,198,Ricardo Ryller,Ricardo,90.000000
6136,986,Pedro Ferreira,Ferreira,90.000000
865,285,Murilo,Murilo Cerqueira,90.000000
3195,651,Ryan,Ryan Gustavo,90.000000
2774,603,Naves,Kaiky Naves,90.000000


In [230]:
# manually reviewed: these matches were rejected as false positives
# (e.g. same club, but different real people)
rejected_player_idx = [440]  # Marcelo Pitaluga (440) incorrectly matched to Marcelo

df_accepted_matches_reviewed = df_accepted_matches[~df_accepted_matches['player_idx'].isin(rejected_player_idx)]
df_accepted_matches_reviewed

,player_idx,name_x,candidate,score
2524,568,Gustavo Martins,Gustavo Martins Santos,95.000000
799,265,José Luis Rodríguez,José Luis Rodríguez Bebanz,95.000000
1876,460,Ewerthon,Ewerton,93.333333
1334,360,Eduardo,Eduardo Santos,90.000000
380,144,João Ricardo,João Ricardo Riedi,90.000000
533,198,Ricardo Ryller,Ricardo,90.000000
6136,986,Pedro Ferreira,Ferreira,90.000000
865,285,Murilo,Murilo Cerqueira,90.000000
3195,651,Ryan,Ryan Gustavo,90.000000
2774,603,Naves,Kaiky Naves,90.000000


In [231]:
matched_stats_names = df_players_brazil[df_players_brazil['name_normalized'].isin(df_stats_by_player['name_normalized'])]['name_normalized']
overlap = df_accepted_matches_reviewed['candidate'].isin(
    df_stats_by_player[df_stats_by_player['name_normalized'].isin(matched_stats_names)]['Jogador']
)
overlap.sum()

np.int64(0)

In [232]:
df_accepted_matches.shape

(31, 4)

In [233]:
df_accepted_matches[['name_x', 'candidate', 'score']].sort_values('score', ascending=False)

,name_x,candidate,score
2524,Gustavo Martins,Gustavo Martins Santos,95.000000
799,José Luis Rodríguez,José Luis Rodríguez Bebanz,95.000000
1876,Ewerthon,Ewerton,93.333333
1334,Eduardo,Eduardo Santos,90.000000
380,João Ricardo,João Ricardo Riedi,90.000000
533,Ricardo Ryller,Ricardo,90.000000
6136,Pedro Ferreira,Ferreira,90.000000
865,Murilo,Murilo Cerqueira,90.000000
3195,Ryan,Ryan Gustavo,90.000000
2774,Naves,Kaiky Naves,90.000000


In [234]:
rejected_names = ['Gabriel Rojas', 'Richard', 'Ricardo Ryller', 'Souza', 'Eduardo']
df_accepted_matches_reviewed = df_accepted_matches[~df_accepted_matches['name_x'].isin(rejected_names)]
df_accepted_matches_reviewed.shape

(25, 4)

## Reviewing remaining unmatched players with candidate suggestions

In [235]:
# players still unmatched after direct match + accepted fuzzy matches
fuzzy_matched_idx = df_accepted_matches_reviewed['player_idx'].tolist()
still_unmatched = no_match[~no_match.index.isin(fuzzy_matched_idx)]
still_unmatched = no_match[~no_match.index.isin(fuzzy_matched_idx)].drop(columns=['fuzzy_candidate', 'fuzzy_score'])
still_unmatched.shape
still_unmatched.shape

(621, 44)

In [236]:
still_unmatched.columns.tolist()

['player_id',
 'first_name',
 'last_name',
 'name_x',
 'last_season_x',
 'current_club_id',
 'player_code',
 'country_of_birth',
 'city_of_birth',
 'country_of_citizenship',
 'date_of_birth',
 'sub_position',
 'position',
 'foot',
 'height_in_cm',
 'contract_expiration_date',
 'agent_name',
 'image_url',
 'international_caps',
 'international_goals',
 'current_national_team_id',
 'url_x',
 'current_club_domestic_competition_id',
 'current_club_name',
 'market_value_in_eur',
 'highest_market_value_in_eur',
 'club_id',
 'club_code',
 'name_y',
 'domestic_competition_id',
 'total_market_value',
 'squad_size',
 'average_age',
 'foreigners_number',
 'foreigners_percentage',
 'national_team_players',
 'stadium_name',
 'stadium_seats',
 'net_transfer_record',
 'coach_name',
 'last_season_y',
 'filename',
 'url_y',
 'name_normalized']

In [237]:
from rapidfuzz import process

review_rows = []
for idx, row in still_unmatched.iterrows():
    name = row['name_x']
    last_name = name.split()[-1]  # take the last word of the name (usually the surname)

    # pre-filter: only consider candidates whose name contains the surname
    candidates_pool = df_stats_by_player[df_stats_by_player['Jogador'].str.contains(last_name, case=False, na=False)]

    if not candidates_pool.empty:
        top_matches = process.extract(name, candidates_pool['Jogador'], scorer=fuzz.WRatio, limit=3)
        review_rows.append({
            'name_x': name,
            'current_club_name': row['current_club_name'],
            'top_candidates': top_matches
        })

df_review = pd.DataFrame(review_rows)
df_review

,name_x,current_club_name,top_candidates
0,Léo Vieira,Esporte Clube Bahia,"[(Dudu Vieira, 71.25, 125), (Felipe Vieira, 71..."
1,Souza,Club de Regatas Vasco da Gama,"[(Cauly Oliveira Souza, 90.0, 81), (Gabriel So..."
2,Oscar,São Paulo Futebol Clube,"[(Oscar Estupiñan, 90.0, 487), (Gabriel Moscar..."
3,Gil,Santos Futebol Clube,"[(Gilberto, 90.0, 209)]"
4,Eduardo,Mirasol Futebol Clube,"[(Carlos Eduardo, 90.0, 78), (Eduardo Melo, 90..."
...,...,...,...
248,Cléber Lucas,S. A. F. Botafogo,"[(Lucas Calegari, 73.07692307692307, 360), (Lu..."
249,Lyan Araujo,Esporte Clube Bahia,"[(Michel Araujo, 67.05882352941177, 466)]"
250,Mosquito,Clube Atlético Mineiro,"[(Gustavo Mosquito, 90.0, 231)]"
251,Kauan Vitor,Esporte Clube Vitória,"[(José Vitor Silva Neves, 85.5, 304), (Vitor H..."


In [238]:
with open('../data/processed/review_candidates.txt', 'w', encoding='utf-8') as f:
    for _, row in df_review.iterrows():
        f.write(f"{row['name_x']} ({row['current_club_name']})\n")
        for candidate_name, score, _ in row['top_candidates']:
            f.write(f"    → {candidate_name} (score: {score:.1f})\n")
        f.write("\n")

print("Saved to data/processed/review_candidates.txt")

Saved to data/processed/review_candidates.txt


## Consolidating all matches

In [239]:
manual_accepted = [
    ('Marçal', 'Fernando Marçal'),
    ('Rafa Silva', 'Rafael Silva'),
    ('Emmanuel Martínez', 'Leandro Emmanuel Martínez'),
    ('Félix Torres', 'Félix Torres Caicedo'),
    ('Igor Vinícius', 'Igor Vinícius de Souza'),
    ('Baralhas', 'Gabriel Baralhas'),
    ('Praxedes', 'Bruno Praxedes'),
    ('Léo Linck', 'Leonardo Linck'),
    ('Zé Vitor', 'José Vitor Silva Neves'),
    ('Alix', 'Alix Vinicius'),
    ('Lacerda', 'Derik Lacerda'),
]

df_manual_matches = pd.DataFrame(manual_accepted, columns=['name_x', 'candidate'])
df_manual_matches

,name_x,candidate
0,Marçal,Fernando Marçal
1,Rafa Silva,Rafael Silva
2,Emmanuel Martínez,Leandro Emmanuel Martínez
3,Félix Torres,Félix Torres Caicedo
4,Igor Vinícius,Igor Vinícius de Souza
5,Baralhas,Gabriel Baralhas
6,Praxedes,Bruno Praxedes
7,Léo Linck,Leonardo Linck
8,Zé Vitor,José Vitor Silva Neves
9,Alix,Alix Vinicius


## Checking for 2024 players no longer in Brazil

In [240]:
# names from df_stats_by_player already claimed by some match
used_direct = df_players_brazil[
    df_players_brazil['name_normalized'].isin(df_stats_by_player['name_normalized'])
]['name_normalized']

used_fuzzy_club = df_accepted_matches_reviewed['candidate']
used_fuzzy_manual = df_manual_matches['candidate']

# mark which df_stats_by_player rows were used, from any of the three sources
df_stats_by_player['used'] = (
    df_stats_by_player['name_normalized'].isin(used_direct) |
    df_stats_by_player['Jogador'].isin(used_fuzzy_club) |
    df_stats_by_player['Jogador'].isin(used_fuzzy_manual)
)

unused_stats_players = df_stats_by_player[~df_stats_by_player['used']]
unused_stats_players.shape

(280, 20)

In [241]:
print("Total stats players:", df_stats_by_player.shape[0])
print("Used via direct match:", df_stats_by_player['name_normalized'].isin(used_direct).sum())
print("Used via fuzzy club match:", df_stats_by_player['Jogador'].isin(used_fuzzy_club).sum())
print("Used via fuzzy manual match:", df_stats_by_player['Jogador'].isin(used_fuzzy_manual).sum())

Total stats players: 652
Used via direct match: 336
Used via fuzzy club match: 25
Used via fuzzy manual match: 11


In [242]:
matched_players = df_players_brazil[df_players_brazil['name_normalized'].isin(df_stats_by_player['name_normalized'])]
matched_players['name_normalized'].duplicated().sum()

np.int64(67)

## Resolving homonym duplicates in direct match

In [243]:
duplicated_names = matched_players[matched_players['name_normalized'].duplicated(keep=False)]
duplicated_names[['name_x', 'current_club_name', 'name_normalized']].sort_values('name_normalized')

,name_x,current_club_name,name_normalized
246,Alan Franco,Clube Atlético Mineiro,alan franco
314,Alan Franco,São Paulo Futebol Clube,alan franco
1043,Alisson,Sport Club Internacional,alisson
120,Alisson,Fluminense Football Club,alisson
54,Allan,S. A. F. Botafogo,allan
...,...,...,...
591,Vitinho,S. A. F. Botafogo,vitinho
292,Vitinho,S. A. F. Botafogo,vitinho
542,Vitinho,Clube Atlético Mineiro,vitinho
928,Wanderson,Esporte Clube Vitória,wanderson


In [244]:
# check for exact duplicate rows (same name AND same club) — likely a data
# quality issue upstream, not genuine homonyms
exact_duplicates = duplicated_names[duplicated_names.duplicated(subset=['name_x', 'current_club_name'], keep=False)]
exact_duplicates.sort_values('name_normalized')

,player_id,first_name,last_name,name_x,last_season_x,current_club_id,player_code,country_of_birth,city_of_birth,country_of_citizenship,...,national_team_players,stadium_name,stadium_seats,net_transfer_record,coach_name,last_season_y,filename,url_y,name_normalized,matched
85,174570,NaN,Gabriel,Gabriel,2024,3876,gabriel,Brazil,Salvador,Brazil,...,0,Estádio Municipal José María de Campos Maia,14534,+€1.67m,NaN,2025,../data/raw/transfermarkt-scraper/2025/clubs.j...,https://www.transfermarkt.co.uk/mirassol-futeb...,gabriel,True
635,995883,NaN,Gabriel,Gabriel,2025,3876,gabriel,Brazil,Guarapuava,Brazil,...,0,Estádio Municipal José María de Campos Maia,14534,+€1.67m,NaN,2025,../data/raw/transfermarkt-scraper/2025/clubs.j...,https://www.transfermarkt.co.uk/mirassol-futeb...,gabriel,True
71,149577,NaN,Pablo,Pablo,2024,614,pablo,Brazil,São Luís,Brazil,...,10,Estádio Jornalista Mário Filho,82238,€-13.92m,NaN,2025,../data/raw/transfermarkt-scraper/2025/clubs.j...,https://www.transfermarkt.co.uk/flamengo-rio-d...,pablo,True
954,1332141,NaN,Pablo,Pablo,2024,614,pablo,Brazil,Joinville,Brazil,...,10,Estádio Jornalista Mário Filho,82238,€-13.92m,NaN,2025,../data/raw/transfermarkt-scraper/2025/clubs.j...,https://www.transfermarkt.co.uk/flamengo-rio-d...,pablo,True
692,1082692,NaN,Paulinho,Paulinho,2025,978,paulinho,Brazil,Ribeirão Preto,Brazil,...,5,Estádio São Januário,24584,+€23.85m,NaN,2025,../data/raw/transfermarkt-scraper/2025/clubs.j...,https://www.transfermarkt.co.uk/vasco-da-gama-...,paulinho,True
716,1083888,NaN,Paulinho,Paulinho,2024,978,paulinho,Brazil,Manaus,Brazil,...,5,Estádio São Januário,24584,+€23.85m,NaN,2025,../data/raw/transfermarkt-scraper/2025/clubs.j...,https://www.transfermarkt.co.uk/vasco-da-gama-...,paulinho,True
332,531916,NaN,Pedro Henrique,Pedro Henrique,2025,2125,pedro-henrique,Brazil,Goiânia,Brazil,...,0,Estádio Manoel Barradas,30793,+€1.36m,NaN,2025,../data/raw/transfermarkt-scraper/2025/clubs.j...,https://www.transfermarkt.co.uk/esporte-clube-...,pedro henrique,True
891,1223830,NaN,Pedro Henrique,Pedro Henrique,2024,2125,pedro-henrique,NaN,NaN,Brazil,...,0,Estádio Manoel Barradas,30793,+€1.36m,NaN,2025,../data/raw/transfermarkt-scraper/2025/clubs.j...,https://www.transfermarkt.co.uk/esporte-clube-...,pedro henrique,True
292,468249,NaN,Vitinho,Vitinho,2025,537,vitinho,Brazil,Belo Horizonte,Brazil,...,5,Estádio Olímpico Nilton Santos,46931,+€40.04m,NaN,2025,../data/raw/transfermarkt-scraper/2025/clubs.j...,https://www.transfermarkt.co.uk/botafogo-rio-d...,vitinho,True
591,928096,NaN,Vitinho,Vitinho,2024,537,vitinho,NaN,NaN,Brazil,...,5,Estádio Olímpico Nilton Santos,46931,+€40.04m,NaN,2025,../data/raw/transfermarkt-scraper/2025/clubs.j...,https://www.transfermarkt.co.uk/botafogo-rio-d...,vitinho,True


In [245]:
# remove unresolvable duplicates: same name AND same club — no way to
# tell the two real players apart with the data available, so both are
# dropped from the direct match to avoid assigning stats to the wrong person
exact_duplicate_names = exact_duplicates['name_normalized'].unique()

df_players_brazil_clean = df_players_brazil[
    ~((df_players_brazil['name_normalized'].isin(exact_duplicate_names)) &
      (df_players_brazil['name_normalized'].isin(matched_players['name_normalized'])))
]
df_players_brazil_clean.shape

(1021, 45)

## Final merge

In [246]:
# direct matches: name_normalized as the join key
direct_matches = df_players_brazil_clean[
    df_players_brazil_clean['name_normalized'].isin(df_stats_by_player['name_normalized'])
][['name_x', 'name_normalized']].copy()
direct_matches['stats_key'] = direct_matches['name_normalized']

# fuzzy matches (club filter + manual): map name_x -> the stats player's name_normalized
fuzzy_club_matches = df_accepted_matches_reviewed[['name_x', 'candidate']].copy()
fuzzy_club_matches['stats_key'] = fuzzy_club_matches['candidate'].apply(normalize_name)

fuzzy_manual_matches = df_manual_matches[['name_x', 'candidate']].copy()
fuzzy_manual_matches['stats_key'] = fuzzy_manual_matches['candidate'].apply(normalize_name)

all_matches = pd.concat([
    direct_matches[['name_x', 'stats_key']],
    fuzzy_club_matches[['name_x', 'stats_key']],
    fuzzy_manual_matches[['name_x', 'stats_key']],
])
all_matches.shape

(411, 2)

In [247]:
duplicated_stats_keys = all_matches[all_matches.duplicated(subset='name_x', keep=False)]['name_x'].unique()
all_matches_clean = all_matches[~all_matches['name_x'].isin(duplicated_stats_keys)]
all_matches_clean.shape

(324, 2)

In [248]:
df_final = df_players_brazil_clean.merge(all_matches_clean, on='name_x', how='left')
df_final = df_final.merge(df_stats_by_player, left_on='stats_key', right_on='name_normalized', how='left', suffixes=('', '_stats'))
df_final.shape

(1022, 66)

In [249]:
df_final['Jogador'].notna().sum()

np.int64(327)

In [250]:
df_final.shape[0], df_players_brazil_clean.shape[0]

(1022, 1021)

In [251]:
all_matches['name_x'].duplicated().sum()

np.int64(46)

In [252]:
all_matches[all_matches['name_x'].duplicated(keep=False)].sort_values('name_x')

,name_x,stats_key
314,Alan Franco,alan franco
246,Alan Franco,alan franco
120,Alisson,alisson
1043,Alisson,alisson
690,Allan,allan
...,...,...
513,Victor Hugo,victor hugo
972,Vitão,vitao
263,Vitão,vitao
928,Wanderson,wanderson


In [253]:
# for homonyms in different clubs, keep only the one whose club matches
# the club registered in the stats dataset (df_stats_by_player['Time'])
def resolve_homonym(group):
    if len(group) == 1:
        return group
    stats_key = group['stats_key'].iloc[0]
    stats_club = df_stats_by_player[df_stats_by_player['name_normalized'] == stats_key]['Time']
    if stats_club.empty:
        return group.iloc[[0]]  # fallback: just keep the first one
    stats_club = stats_club.iloc[0]
    # find which player in the group has a club name matching (or contained in) the stats club
    for idx, row in group.iterrows():
        player_club = df_players_brazil_clean[df_players_brazil_clean['name_x'] == row['name_x']]['current_club_name']
        # this is imperfect when name_x repeats across different real players — see note below
    return group.iloc[[0]]  # placeholder, see explanation below

all_matches_deduped = all_matches.drop_duplicates(subset='name_x', keep='first')
all_matches_deduped.shape

(365, 2)

In [254]:
df_final['name_x'].value_counts().head(3)

name_x
Riquelme    4
Eduardo     3
Allan       3
Name: count, dtype: int64

In [255]:
df_final['player_id'].duplicated().sum()

np.int64(1)

In [256]:
duplicated_id = df_final[df_final['player_id'].duplicated(keep=False)]
duplicated_id[['player_id', 'name_x', 'current_club_name', 'Jogador', 'Time']]

,player_id,name_x,current_club_name,Jogador,Time
20,69400,Weverton,Grêmio Foot-Ball Porto Alegrense,Weverton,Cruzeiro
21,69400,Weverton,Grêmio Foot-Ball Porto Alegrense,Wéverton,Palmeiras


In [257]:
df_final_clean = df_final[~((df_final['player_id'] == 69400) & (df_final['Jogador'] == 'Weverton'))].copy()
df_final_clean.shape

(1021, 66)

In [258]:
df_final_clean['Jogador'].notna().sum()

np.int64(326)

In [259]:
print("all_matches (raw):", all_matches.shape[0])
print("all_matches unique name_x:", all_matches['name_x'].nunique())
print("all_matches_clean:", all_matches_clean.shape[0])

all_matches (raw): 411
all_matches unique name_x: 365
all_matches_clean: 324


In [260]:
df_final_clean.to_csv('../data/processed/players_with_stats_2024.csv', index=False)

## Conclusion

- Merged df_players_brazil (Transfermarkt, 1021 players after removing
  unresolvable same-name-same-club duplicates) with performance statistics
  from a secondary source (Eduardo Palmieri, Série A 2024), matched by name
  since no shared ID exists between the two sources.
- Matching combined three techniques: direct normalized-name match (401),
  fuzzy matching with club as a blocking filter and score >= 81 (25,
  manually reviewed), and manual fuzzy matches without club filter to
  catch players who changed clubs between 2024 and now (11, manually
  reviewed).
- Resolved duplicate/ambiguous matches: homonyms with identical name AND
  club were dropped (unresolvable), homonyms matching the same stats
  candidate were dropped from the match set, and one specific case
  (Weverton, Cruzeiro vs Palmeiras) was manually resolved after confirming
  the correct club with the user.
- - Final result: 1021 players in the 2024 Série A roster, with ~326 having
  matched performance statistics (goals, assists, xG, xAG, progressive
  passes/carries, defensive actions). The rest are unmatched — some because
  they didn't play in 2024 (reserves, youth, or players who joined later),
  and some due to unresolved name ambiguity (homonyms across different
  clubs, conservatively excluded rather than risking a wrong assignment).
- Known limitation: name-based matching, even with careful validation, may
  contain a small number of undetected errors (swapped homonyms, missed
  transfers). This should be treated as reasonably reliable but not
  perfectly clean data.
- Result exported to data/processed/players_with_stats_2024.csv, ready for
  the Fase 1 cost-benefit scoring and clustering steps.